provider/oracle/network/environ.go

```diff
@@ -0,0 +1,347 @@
+type Environ struct {
+	client NetworkingAPI
+}
+
+var _ environs.Networking = (*Environ)(nil)
+
+// NewEnviron returns a new network Environment that has
+// network capabilities inside the oracle cloud environment
+func NewEnviron(api NetworkingAPI) *Environ {
+	return &Environ{
+		client: api,
+	}
+}
+
+// Subnets returns basic information about subnets known by the provider
+// for the oracle cloud environment
+func (e Environ) Subnets(
+	id instance.Id,
+	subnets []network.Id,
+) ([]network.SubnetInfo, error) {
+
+	ret := []network.SubnetInfo{}
+	found := make(map[string]bool)
+	if id != instance.UnknownId {
+		instanceNets, err := e.NetworkInterfaces(id)
+		if err != nil {
+			return ret, errors.Trace(err)
+		}
+		if len(subnets) == 0 {
+			for _, val := range instanceNets {
+				found[string(val.ProviderSubnetId)] = false
+			}
+		}
+		for _, val := range instanceNets {
+			if _, ok := found[string(val.ProviderSubnetId)]; !ok {
+				continue
+			} else {
+				found[string(val.ProviderSubnetId)] = true
+				subnetInfo := network.SubnetInfo{
+					CIDR:            val.CIDR,
+					ProviderId:      val.ProviderSubnetId,
+					SpaceProviderId: val.ProviderSpaceId,
+				}
+				ret = append(ret, subnetInfo)
+			}
+		}
+	} else {
+		subnets, err := e.getSubnetInfoAsMap()
+		if err != nil {
+			return ret, errors.Trace(err)
+		}
+		if len(subnets) == 0 {
+			for key, _ := range subnets {
+				found[key] = false
+			}
+		}
+		for key, val := range subnets {
+			if _, ok := found[key]; !ok {
+				continue
+			}
+			found[key] = true
+			ret = append(ret, val)
+		}
+	}
+	missing := []string{}
+	for key, ok := range found {
+		if !ok {
+			missing = append(missing, key)
+		}
+	}
+	if len(missing) > 0 {
+		return nil, errors.Errorf(
+			"missing subnets: %s", strings.Join(missing, ","))
+	}
+	return ret, nil
+}
+
+// getSubnetInfoAsMap will return the subnet information
+// for the getSubnetInfo as a map rathan than a slice
+func (e Environ) getSubnetInfoAsMap() (map[string]network.SubnetInfo, error) {
+	subnets, err := e.getSubnetInfo()
+	if err != nil {
+		return nil, err
+	}
+	ret := make(map[string]network.SubnetInfo, len(subnets))
+	for _, val := range subnets {
+		ret[string(val.ProviderId)] = val
+	}
+	return ret, nil
+}
+
+// getSubnetInfo will return the subnet information of the current oracle cloud
+// environ. This will query all oracle ip networks currenty unde the indentity
+// domain name. The result of the query will be parsed as a slice of
+// network.SubnetInfo
+func (e Environ) getSubnetInfo() ([]network.SubnetInfo, error) {
+	networks, err := e.client.AllIpNetworks(nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	subnets := make([]network.SubnetInfo, len(networks.Result))
+	idx := 0
+	for _, val := range networks.Result {
+		var spaceId network.Id
+		if val.IpNetworkExchange != nil {
+			spaceId = network.Id(*val.IpNetworkExchange)
+		}
+		subnets[idx] = network.SubnetInfo{
```

Comment:
In a nutshell, efficiency. A slice is simply a subset of an array. Both slices and arrays are of fixed size. When you use ```append()```, behind the scenes, a new slice is created of ```len(originalSlice) + len(valuesToadd)```. Values from the original slice are copied over to the new one, and ```valuesToAdd``` are added at the end of the new slice.

While using ```append()``` easier to read, I find that the above approach is not that terrible to warrant using ```append()```. If you prefer I can change it.